anonlink/serialization.py

```diff
@@ -0,0 +1,221 @@
+import array as _array
+import heapq as _heapq
+import io as _io
+import struct as _struct
+import typing as _typing
+
+from . import typechecking as _typechecking
+
+# https://docs.python.org/3/library/struct.html#format-characters
+_STRUCT_UINT_LEN_TO_FMT = {1: 'B', 2: 'H', 4: 'L', 8: 'Q'}
+_STRUCT_FLOAT_LEN_TO_FMT = {2: 'e', 4: 'f', 8: 'd'}
+
+# https://docs.python.org/3/library/array.html#module-array
+_ARRAY_UINT_TYPES = 'BIHLQ'
+_ARRAY_FLOAT_TYPES = 'df'
+_ARRAY_UINT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                          for t in _ARRAY_UINT_TYPES}
+_ARRAY_FLOAT_LEN_TO_FMT = {_array.array(t).itemsize: t
+                           for t in _ARRAY_FLOAT_TYPES}
+
+_HEADER_STRUCT = _struct.Struct('<BBBB')
+
+
+_CandidatePair = _typing.Tuple[float, int, int, int, int]
+_CandidatePairIter = _typing.Iterable[_CandidatePair]
+
+# This is subject to change.
+# The format is composed of a header and a sequence of entries.
+# The header is composed of:  version (1 byte), number of bytes in
+# similarity (1 byte), number of bytes in dataset index (1 byte), number
+# of bytes in record index (1 byte).
+# Every entry is composed of a similarity as a IEEE floating point
+# value, a two datset indices, and two record indices. Their sizes as
+# determined by the header.
+# All integers are unsigned. All values are *little-endian* (yes, this
+# breaks convention). The number of candidate matchings is not included
+# on purpose: if needed, it can be computed from the length of the file.
+
+
+def _entry_struct(
+    sim_t_size: int,
+    dset_i_t_size: int,
+    rec_i_t_size: int
+) -> _struct.Struct:
+    try:
+        sim_t = _STRUCT_FLOAT_LEN_TO_FMT[sim_t_size]
+    except KeyError:
+        msg = f'floats of {sim_t_size} bytes are not supported'
+        raise ValueError(msg) from None
```

Comment:
Does the `from None` do anything useful?

Reply:
It does! It’s part of [PEP 409](https://www.python.org/dev/peps/pep-0409/). It gets rid of the exception context. In other words, when we raise `ValueError`, we forget about `KeyError`.

The `KeyError` is just an unimportant implementation detail. This prevents the user being told that ‘`During handling of the above exception, another exception occurred`’.

Reply:
Cheers, I'll try remember that!